numba/tests/test_array_methods.py

```diff
@@ -779,18 +787,198 @@ def test_sum(self):
+        # OK
+        a = np.ones((7, 6, 5, 4, 3))
+        self.assertFalse(type(pyfunc(a, dtype=np.int32)) == cfunc(a, dtype=dtype))
+
+    def test_sum_axis_dtype_kws(self):
+        pyfunc = array_sum_axis_dtype_kws
+        cfunc = jit(nopython=True)(pyfunc)
+        dtype = np.float64
+        # OK
+        a = np.ones((7, 6, 5, 4, 3))
+        self.assertPreciseEqual(pyfunc(a, axis=1, dtype=dtype),
+                                cfunc(a,  axis=1, dtype=dtype))
+        self.assertPreciseEqual(pyfunc(a, axis=2, dtype=dtype),
+                                cfunc(a, axis=2, dtype=dtype))
+
+    def test_sum_axis_dtype_kws2(self):
+        pyfunc = array_sum_axis_dtype_kws
+        cfunc = jit(nopython=True)(pyfunc)
+        all_dtypes = [np.float64, np.float32, np.int64, np.int32, np.uint32,
+                      np.uint64, np.complex64, np.complex128]
+        all_test_arrays = [
+            [np.ones((7, 6, 5, 4, 3), arr_dtype),
+             np.ones(1, arr_dtype),
+             np.ones((7, 3), arr_dtype) * -5]
+            for arr_dtype in all_dtypes]
+
+        out_dtypes = {np.dtype('float64'): [np.float64],
+                      np.dtype('float32'): [np.float64, np.float32],
+                      np.dtype('int64'): [np.float64, np.int64, np.float32],
+                      np.dtype('int32'): [np.float64, np.int64, np.float32, np.int32],
+                      np.dtype('uint32'): [np.float64, np.int64, np.float32],
+                      np.dtype('uint64'): [np.float64, np.uint64],
+                      np.dtype('complex64'): [np.complex64],
+                      np.dtype('complex128'): [np.complex128],
+                      np.dtype('timedelta64'): [np.timedelta64]}
+
+        for arr_list in all_test_arrays:
+            for arr in arr_list:
+                for out_dtype in out_dtypes[arr.dtype]:
+                    for axis in (0, 1, 2):
+                        if axis > len(arr.shape) - 1:
+                            continue
+                        subtest_str = ("Testing np.sum with {} input and {} output "
+                                       .format(arr.dtype, out_dtype))
+                        with self.subTest(subtest_str):
+                            py_res = pyfunc(arr, axis=axis, dtype=out_dtype)
+                            nb_res = cfunc(arr, axis=axis, dtype=out_dtype)
+                            self.assertPreciseEqual(py_res, nb_res)
+
+    def test_sum_axis_dtype_pos_arg(self):
+        """
+        testing that axis and dtype inputs work when passed as positional
+        :return:
+        """
+        pyfunc = array_sum_axis_dtype_pos
+        cfunc = jit(nopython=True)(pyfunc)
+        dtype = np.float64
+        # OK
+        a = np.ones((7, 6, 5, 4, 3))
+        self.assertPreciseEqual(pyfunc(a, 1, dtype),
+                                cfunc(a,  1, dtype))
+
+        self.assertPreciseEqual(pyfunc(a, 2, dtype),
+                                cfunc(a, 2, dtype))
+
```

Comment:
This seems to not work:
```python
from numba import njit
import numpy as np

@njit
def foo(a, dtype):
    z =  a.sum(dtype=dtype, axis=1)
    return z


a = np.ones((2, 3), dtype=np.int8)
print(foo(a, np.uint32))
```

Reply:
I'm actually wondering if this should work at all under Numba semantics for `+=`?

Reply:
> This seems to not work:

it's rejecting the transformation of signed integer into unsigned integer. isn't that ok?

> I'm actually wondering if this should work at all under Numba semantics for `+=`?

I didn't understand why `+=` would play a role here.

Reply:
I think we can just leave it for now, the complexity of untangling this outweighs the benefit given it will be rewritten/refactored in the near future. The `+=` is to do with reducing where the accumulator is an array https://github.com/numba/numba/pull/4472/files#diff-4b9c6924f184d92f076cc50b84bcc442R224